### PR TITLE
NMS-13118/NMS-13109: Modified handling of meta-data

### PR DESCRIPTION
**Pull request by christianpape**
* JIRA: https://issues.opennms.org/browse/NMS-13118
* JIRA: https://issues.opennms.org/browse/NMS-13109

### CODE REVIEW

**Review by cgorantla — CHANGES_REQUESTED** [review]

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java [hunk 1]
```diff
@@ -258,12 +258,30 @@ public void updateNode(final OnmsNode node, String rescanExisting) {
...
-                node.getIpInterfaceByIpAddress(inetAddress).setMetaData(onmsIpInterface.getMetaData());
+                final OnmsIpInterface nodeInterface = node.getIpInterfaceByIpAddress(inetAddress);
+                if (nodeInterface != null) {
+                    for(final OnmsMetaData onmsMetaData : dbNodeInterface.getMetaData()) {
+                        if (!nodeInterface.findMetaDataForContextAndKey(onmsMetaData.getContext(), onmsMetaData.getKey()).isPresent()) {
+                            nodeInterface.addMetaData(onmsMetaData.getContext(), onmsMetaData.getKey(), onmsMetaData.getValue());
+                        }
+                    }
+                    for (OnmsMonitoredService dbNodeMonitoredService:dbNodeInterface.getMonitoredServices()) {
+                        final OnmsMonitoredService nodeMonitoredService = nodeInterface.getMonitoredServiceByServiceType(dbNodeMonitoredService.getServiceName());
```
[the remark this answers is not in the document]
Comment (christianpape) [reply opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java:277]: done.

**Comment by indigo423** [comment]
Here is what I've tested:

## NMS-13118
* Added a new node in a requisition and synched it with "Yes"  -> 👍 
* Changed meta-data and added a new attribute and synched with "Yes" and "DB only" -> 👍 

## NMS-13109
* Imported nodes from vCenter with Provisiond schedule and added additional meta-data manually -> 👍 
* Reimported vCenter with Provisiond schedule and check if meta-data is still available -> 👍

**Review by indigo423 — APPROVED** [review]
Here is what I've tested:

## NMS-13118
* Added a new node in a requisition and synched it with "Yes"  -> 👍 
* Changed meta-data and added a new attribute and synched with "Yes" and "DB only" -> 👍 

## NMS-13109
* Imported nodes from vCenter with Provisiond schedule and added additional meta-data manually -> 👍 
* Reimported vCenter with Provisiond schedule and check if meta-data is still available -> 👍

File: opennms-provision/opennms-provisiond/src/test/java/org/opennms/netmgt/provision/service/MetaDataIT.java [hunk 2]
```diff
@@ -191,4 +196,83 @@ public void testScanSpaceDevice() throws Exception {
...
+        assertThat(node.getIpInterfaceByIpAddress("192.168.3.1").getMetaData().size(), is(1));
+        assertThat(node.getIpInterfaceByIpAddress("192.168.3.1").getMetaData(), containsInAnyOrder(
+                new OnmsMetaData("interfaceCustomContext3","interfaceKey3","interfaceValue3")
+        ));
+
+        // now add custom meta-data, see NMS-13109
+        node.addMetaData("nodeCustomContext3", "nodeKey3", "nodeValue3");
+        node.addMetaData("nodeCustomContext1", "nodeKey1", "modifiedNodeValue1"); // this should be overwritten on next sync
+        node.getIpInterfaceByIpAddress("192.168.3.1").getMetaData().add(new OnmsMetaData("interfaceCustomContext4", "interfaceKey4", "interfaceValue4"));
+        m_nodeDao.update(node);
```
Comment (cgorantla) [inline opennms-provision/opennms-provisiond/src/test/java/org/opennms/netmgt/provision/service/MetaDataIT.java:239]: Can we add metadata for Service as well and verify that it updates properly.

**Review by cgorantla — APPROVED** [review]

**Pull request merged**